src/ring_theory/derivation.lean

```diff
@@ -0,0 +1,223 @@
+
+TODO: this file is just a stub to go on with some PRs in the geometry section. It only
+implements the definition of derivations in commutative algebra. This will soon change: as soon
+as bimodules will be there in mathlib I will change this file to take into account the
+non-commutative case. Any development on the theory of derivations is discouraged until the
+definitive definition of derivation will be implemented.
+
+Also note that this file implements bundled derivations only by now. To be better integrated with
+the algebra section of mathlib it would probably be better to define a notion of `is_derivation` and
+link it to bundled derivations.
+-/
+
+open algebra ring_hom
+
+instance algebra.to_is_scalar_tower  {R : Type*} {A : Type*} [comm_semiring R] [semiring A]
+  [algebra R A] : is_scalar_tower R A A :=
+⟨λ r a b, by simp only [smul_def]; exact mul_smul _ _ _⟩
+
+section is_scalar_tower
+
+variables {R : Type*} [comm_semiring R]
+variables (A : Type*) [comm_semiring A] [algebra R A]
+variables {M : Type*} [add_comm_monoid M] [semimodule A M] [semimodule R M]
+  [is_scalar_tower R A M]
+variables {N : Type*} [add_comm_monoid N] [semimodule A N] [semimodule R N] [is_scalar_tower R A N]
+
+@[simp] lemma algebra_compatible_smul (r : R) (m : M) : r • m = ((algebra_map R A) r) • m :=
+by rw [←(one_smul A m), ←smul_assoc, algebra.smul_def, mul_one, one_smul]
+
+variable {A}
+
+@[simp] lemma algebra_compatible_smul_comm (r : R) (a : A) (m : M) : a • r • m = r • a • m :=
+by rw [algebra_compatible_smul A r m, algebra_compatible_smul A r (a • m), ←mul_smul, mul_comm, mul_smul]
+
+@[simp] lemma compatible_map_smul (f : M →ₗ[A] N) (r : R) (m : M) :
+  f (r • m) = r • f m :=
+by rw [algebra_compatible_smul A r m, linear_map.map_smul, ←algebra_compatible_smul A r (f m)]
+
+instance : has_coe (M →ₗ[A] N) (M →ₗ[R] N) :=
+⟨λ f, ⟨f.to_fun, λ x y, f.map_add' x y, λ r n, compatible_map_smul _ _ _⟩⟩
+
+end is_scalar_tower
+
+/-- `D : derivation R A M` is an `R`-linear map from `A` to `M` that satisfies the `leibniz`
+equality.
+TODO: update this when bimodules are defined. -/
+@[protect_proj]
+structure derivation (R : Type*) (A : Type*) [comm_semiring R] [comm_semiring A]
+  [algebra R A] (M : Type*) [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+  [is_scalar_tower R A M]
+  extends A →ₗ[R] M :=
+(leibniz' (a b : A) : to_fun (a * b) = a • to_fun b + b • to_fun a)
+
+namespace derivation
+
+section
+
+variables {R : Type*} [comm_semiring R]
+{A : Type*} [comm_semiring A] [algebra R A]
+{M : Type*} [add_cancel_comm_monoid M] [semimodule A M] [semimodule R M]
+[is_scalar_tower R A M]
+(D : derivation R A M) {D1 D2 : derivation R A M} (r : R) (a b : A)
+
+instance : has_coe_to_fun (derivation R A M) := ⟨_, λ D, D.to_linear_map.to_fun⟩
+
+instance has_coe_to_linear_map : has_coe (derivation R A M) (A →ₗ[R] M) :=
+  ⟨λ D, D.to_linear_map⟩
+
+@[simp] lemma to_fun_eq_coe : D.to_fun = ⇑D := rfl
```

Comment:
I'd like to disambiguate between `coe_fn D : A -> M` and `coe : A ->[R] M`:
```suggestion
@[simp] lemma to_fun_eq_coe_fn : D.to_fun = ⇑D := rfl
```

Reply:
I am not sure why you think there's ambiguity, It looks like `to_fun_eq_coe` already contains `fun` otherwise it would have been `to_linear_map_eq_coe`. I mean that on the other side of the equal you also get a function you have it for free by how lean works. But most importantly in other parts of mathlib it is adopted the convention I used, that's why I chose this name, so it'd be better to change it everywhere if you believe it needs to be changed.